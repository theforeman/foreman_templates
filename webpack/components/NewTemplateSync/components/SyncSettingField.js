import React from 'react';
import PropTypes from 'prop-types';
import { FieldLevelHelp } from 'patternfly-react';

import TextButtonField from './TextButtonField';
import ButtonTooltip from './ButtonTooltip';

const SyncSettingField = ({ setting, resetField, disabled, syncType }) => {
  const label = settingObj => `${settingObj.fullName} `;

  const fieldSelector = settingObj => {
    if (settingObj.settingsType === 'boolean') {
      return 'checkbox';
    }

    if (settingObj.selection.length !== 0) {
      return 'select';
    }

    return 'text';
  };

  const tooltipContent = (
    <div
      dangerouslySetInnerHTML={{
        __html: setting.description,
      }}
    />
  );

  return (
    <TextButtonField
      name={`${syncType}.${setting.name}`}
      label={label(setting)}
      blank={{}}
      item={setting}
      buttonText={<ButtonTooltip tooltipId={setting.name} />}
      buttonAction={resetField(`${syncType}.${setting.name}`, setting.value)}
      fieldSelector={fieldSelector}
      disabled={disabled}
      fieldRequired={setting.required}
      tooltipHelp={<FieldLevelHelp content={tooltipContent} />}
    >
      {setting.value}
    </TextButtonField>
  );
};

SyncSettingField.propTypes = {
  setting: PropTypes.object.isRequired,
  resetField: PropTypes.func.isRequired,
  disabled: PropTypes.bool.isRequired,
  syncType: PropTypes.string.isRequired,
};

export default SyncSettingField;
