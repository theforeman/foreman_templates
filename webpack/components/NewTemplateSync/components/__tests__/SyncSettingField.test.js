import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import SyncSettingField from '../SyncSettingField';

import {
  associateSetting,
  forceSetting,
  filterSetting,
} from '../../__fixtures__/templateSyncSettings.fixtures';

const noop = () => () => {};

const commonFixtures = {
  resetField: noop,
  disabled: false,
  syncType: 'import',
};

const fixtures = {
  'should render setting with select choices': {
    setting: associateSetting,
    ...commonFixtures,
  },
  'should render setting with input field': {
    setting: filterSetting,
    ...commonFixtures,
  },
  'should render boolean setting as checkbox': {
    setting: forceSetting,
    ...commonFixtures,
  },
};

describe('SyncSettingField', () =>
  testComponentSnapshotsWithFixtures(SyncSettingField, fixtures));
