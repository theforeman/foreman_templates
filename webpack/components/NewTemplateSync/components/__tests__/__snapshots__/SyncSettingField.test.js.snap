// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncSettingField should render boolean setting as checkbox 1`] = `
<TextButtonField
  blank={Object {}}
  buttonAction={[Function]}
  buttonText={
    <ButtonTooltip
      tooltipId="force"
    />
  }
  className=""
  disabled={false}
  fieldRequired={false}
  fieldSelector={[Function]}
  inputClassName="col-md-6"
  item={
    Object {
      "description": "Should importing overwrite locked templates?",
      "id": 46,
      "name": "force",
      "settingsType": "bool",
      "value": false,
    }
  }
  label="undefined "
  name="import.force"
  tooltipHelp={
    <FieldLevelHelp
      buttonClass=""
      content={
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "Should importing overwrite locked templates?",
            }
          }
        />
      }
      placement="top"
      rootClose={true}
    />
  }
/>
`;

exports[`SyncSettingField should render setting with input field 1`] = `
<TextButtonField
  blank={Object {}}
  buttonAction={[Function]}
  buttonText={
    <ButtonTooltip
      tooltipId="filter"
    />
  }
  className=""
  disabled={false}
  fieldRequired={false}
  fieldSelector={[Function]}
  inputClassName="col-md-6"
  item={
    Object {
      "description": "Import or export names matching this regex (case-insensitive; snippets are not filtered)",
      "id": 47,
      "name": "filter",
      "settingsType": "string",
      "value": "",
    }
  }
  label="undefined "
  name="import.filter"
  tooltipHelp={
    <FieldLevelHelp
      buttonClass=""
      content={
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "Import or export names matching this regex (case-insensitive; snippets are not filtered)",
            }
          }
        />
      }
      placement="top"
      rootClose={true}
    />
  }
/>
`;

exports[`SyncSettingField should render setting with select choices 1`] = `
<TextButtonField
  blank={Object {}}
  buttonAction={[Function]}
  buttonText={
    <ButtonTooltip
      tooltipId="associate"
    />
  }
  className=""
  disabled={false}
  fieldRequired={false}
  fieldSelector={[Function]}
  inputClassName="col-md-6"
  item={
    Object {
      "description": "Associate templates to OS, organization and location",
      "id": 45,
      "name": "associate",
      "selection": Array [
        Object {
          "label": "New",
          "value": "new",
        },
        Object {
          "label": "Never",
          "value": "never",
        },
        Object {
          "label": "Always",
          "value": "always",
        },
      ],
      "settingsType": "string",
      "value": "new",
    }
  }
  label="undefined "
  name="import.associate"
  tooltipHelp={
    <FieldLevelHelp
      buttonClass=""
      content={
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "Associate templates to OS, organization and location",
            }
          }
        />
      }
      placement="top"
      rootClose={true}
    />
  }
>
  new
</TextButtonField>
`;
